### PR TITLE
Search Block - ability to specify additional search query vars

### LINKED ISSUE #40583: Search Block extensibility: search query vars
## What problem does this address?

It should be possible to easily create custom search blocks as **core/search** block [variations](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-variations/) with the ability to control [public search query variables](https://codex.wordpress.org/WordPress_Query_Vars#Query_variables) and ability to provide custom taxonomy query vars using the block attributes.

Currently, **core/search** block is already stylable with Global Styles and themes which makes it a great candidate to re-use for different purposes but it is not fully usable as a base for variations because it is not possible to control query parameters using block attributes. 

If we had the ability to specify query parameters as attributes it could enable us to easily create basic search controls for WordPress websites sections or contexts (FAQ/knowledge base pages, search by category, tag or author etc.) or for websites/apps based on extensions that introduce new post types or taxonomies like WooCommerce and products, product categories and product tags.

This is a follow up to a proposal made on [#core-editor](https://wordpress.slack.com/archives/C02QB2JS7/p1649861214648439) Slack weekly meeting. Other mentioned updates to make Search Block more extensible will come as separate feature requests.

 

## What is your proposed solution?

The proposed solution is following patterns that already exist in WordPress with blocks like Query Loop block ([core/query](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/query/block.json)). On the `block.json`, we specify a `query` attribute of the `object` type that takes different query parameters and passes them to the query. For the purpose of the Search Block and the ability to support custom taxonomies following this pattern will work much better than spreading query parameters on the block's attributes like in the Latest Posts block ([core/latest-posts](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/latest-posts/block.json)) as it wouldn't make it possible to support custom taxonomies.

Example core/search variation for WooCommerce product search limited to accessories product category:
```html
<!-- wp:search {"label":"Search","placeholder":"Search products…","buttonText":"Search","query":{"post_type":"product","product_cat":"accessories"}} /-->
```

Based on the populated query object entries we would generate hidden input fields with `key` as input `name` and `value` as `value` that would get automatically included in the search query.

```html
<div class="wp-block-search__inside-wrapper ">
   <input type="search" id="wp-block-search__input-2" class="wp-block-search__input " name="s" value="" placeholder="Search products…" required="">
   <input type="hidden" name="post_type" value="product">
   <input type="hidden" name="product_cat" value="accessories">
   <button type="submit" class="wp-block-search__button  ">Search</button>
</div>
```

I'd suggest splitting design work on Search Block editor UI controls for the query attribute from the implementation as it would already enable us to create block variations using the core/search block and would give us more time to think about which public query vars to expose in the UI and how to best handle custom taxonomies while providing consistent UX for things like auto-complete etc.

We already have [an implementation for query controls](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/query-controls) but it's not really that generic or doesn't provide freedom as to which options we want to expose nor a simple way to provide autocomplete for custom taxonomies etc.
  
 cc: @jeffpaul

### PR DESCRIPTION
**Pull request by tomasztunik**
## What?

Add ability to control search query vars of the core/search block.

Fixes #40583 

## Why?

This will enable creating core/search block variations that can limit their search results based on query vars provided. Ie. searching only among `post_type=product` for WooCommerce or searching only through a particular category of posts, or a particular author on a WordPress page.

## How?

Adds `query` attribute to core/search block and renders hidden inputs in the search block with provided query vars and values. 

## Testing Instructions

1. register new core/search block variation with query attributes

ie. to search only for pages specify `post_type: 'page'` if WooCommerce is available one could use `post_type: 'product'` - it is possible to use all [public query vars](https://codex.wordpress.org/WordPress_Query_Vars#Query_variables) and also custom taxonomies where custom taxonomy name slug is query key and value will be a slug representing custom taxonomy entry.
```
registerBlockVariation( 'core/search', {
   name: 'test/custom-search',
   title: 'Test Search Block',
   category: 'test',
   attributes: {
      query: {
         post_type: 'page'
      }
   }
} );
```

2. Add Test Search Block with inserter
3. Test search and confirm that post_type query parameter is included in browser address bar and that results are limited to given query.

### CODE REVIEW

**Comment by tomasztunik** [comment]
Some questions/ideas I have here.

- Should we allow providing raw query param names as query object keys or should we enter them as `camelCase` and convert them to `snake_case` when creating inputs? Thinking if this could possibly be problematic with custom taxonomies or some custom slug names.
- Should we provide any defaults? we could provide `null` values for empty to expose suggested values for query vars and then don't render inputs if there is no value provided. Thinking if this wouldn't pollute the comment strings for block instances on pages/posts/templates.

**Comment by tomasztunik** [comment]
Just FYI - opened PR to `@types/wordpress__blocks` to include variations registration API which is required to make the best out of this feature before we introduce the UI (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60059)

File: packages/block-library/src/search/edit.js [hunk 1]
```diff
@@ -216,22 +217,33 @@ export default function SearchEdit( {
...
+				/>
+				{ !! query &&
+					Object.keys( query ).map( ( key ) => (
+						<input
+							key={ key }
+							type="hidden"
+							name={ key }
+							value={ query[ key ] }
+						/>
+					) ) }
```
Comment (Mamaduka) [inline packages/block-library/src/search/edit.js]: Hi, @tomasztunik

Sorry if I'm missing something, but what's the reason for outputting `hidden` inputs in the editor?

**Comment by adamziel** [comment]
In a follow-up PR it could be worth exploring editor controls for this.

File: packages/block-library/src/search/edit.js (hunk 1 above)
Comment (tomasztunik) [reply packages/block-library/src/search/edit.js]: Added for consistency in the DOM between the editor/front-end but can be easily removed if we don't see a reason for them to be here.

Comment (Mamaduka) [reply packages/block-library/src/search/edit.js]: Thanks for the explanation, @tomasztunik 🙇 

The hidden inputs shouldn't affect anything visually; they also serve no purpose in the editor besides consistency. So it's probably safe to remove.

**Comment by tomasztunik** [comment]
Agreed, in the [issue](https://github.com/WordPress/gutenberg/issues/40583) I've shared my findings that there already is [query-controls](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/query-controls) component but as mentioned there:
> it's not really that generic or doesn't provide freedom as to which options we want to expose nor a simple way to provide autocomplete for custom taxonomies etc.

Having this first would enable us to create variations already and allow more time for the interface to mature at its own pace

**Comment by Mamaduka** [comment]
> Having this first would enable us to create variations already and allow more time for the interface to mature at its own pace.

Good point 👍

File: packages/block-library/src/search/edit.js (hunk 1 above)
Comment (tomasztunik) [reply packages/block-library/src/search/edit.js]: I had something ringing the bell but double checked and they are completely ignored in rendering, for CSS selectors etc. Will remove them 👍

File: packages/block-library/src/search/edit.js [hunk 2]
```diff
@@ -216,22 +216,24 @@ export default function SearchEdit( {
...
-				// which is confusing.
-				placeholder={
-					placeholder ? undefined : __( 'Optional placeholder…' )
-				}
-				value={ placeholder }
-				onChange={ ( event ) =>
-					setAttributes( { placeholder: event.target.value } )
-				}
-			/>
+			<>
```
Comment (Mamaduka) [inline packages/block-library/src/search/edit.js]: Small nitpick: We can remove Fragment as well.

**Review by Mamaduka — APPROVED** [review]
This looks good to me 👍  Thanks, @tomasztunik!

File: packages/block-library/src/search/edit.js (hunk 2 above)
Comment (tomasztunik) [reply packages/block-library/src/search/edit.js]: 🤦  missed it :)

**Comment by tomasztunik** [comment]
@Mamaduka and @adamziel - thanks for the quick review 🙇

**Comment by mburridge** [comment]
Added Needs Dev Note label in case this needs a dev note for 6.1 release.

**Pull request merged**